### PR TITLE
References types by unique identifier (GroupVersionKind) instead of just Kind throughout Istio

### LINKED ISSUE #20208: Support multiple CRDs with the same type internally
Kubernetes is adding a "Gateway" object. We reference our CRDs by the type (eg "Gateway") in about 10000 places in our codebase. We need to update these to reference by GroupVersionKind (or at least GroupKind)

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/20208

This is not just because I like refactoring, we need this to support new K8s APIs which has a "Gateway".

Since this PR is 1000s of lines of junk, here is the core of it:
* Add a new GroupVersionKind to schema. The name/format is the same as the k8s type, but we copy it to avoid depending on k8s in our core data model
* Config store now does all operations on GroupVersionKind instead of Kind
* Event handlers no do all operations on GroupVersionKind instead of Kind
* Fix a bunch of tests where we only specify Kind, not the full info
* Replace canonical name with GroupVersionKind.String() to avoid duplication
* Completely remove FindByKind() since it isn't safe. Replace with FindByGroupVersionKind()
* Update tons of references to Kind instead of GroupVersionKind

### CODE REVIEW

File: galley/pkg/config/schema/resource/schema.go [hunk 1]
```diff
@@ -74,6 +74,21 @@ type Schema interface {
...
+type GroupVersionKind struct {
+	Group   string
+	Version string
+	Kind    string
+}
+
+var _ fmt.Stringer = GroupVersionKind{}
+
+func (g GroupVersionKind) String() string {
+	if g.Group == "" {
```
Comment (hzxuzhonghu) [inline galley/pkg/config/schema/resource/schema.go]: This can not happen in istio

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -623,7 +624,7 @@ func (s *Server) initEventHandlers() error {
 		pushReq := &model.PushRequest{
 			Full:               true,
 			NamespacesUpdated:  map[string]struct{}{svc.Attributes.Namespace: {}},
-			ConfigTypesUpdated: map[string]struct{}{collections.IstioNetworkingV1Alpha3Serviceentries.Resource().Kind(): {}},
+			ConfigTypesUpdated: map[resource.GroupVersionKind]struct{}{collections.IstioNetworkingV1Alpha3Serviceentries.Resource().GroupVersionKind(): {}},
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go]: Do we really need to care about group version? We have no same resource resides in different groups even when we update version/

File: pilot/pkg/config/kube/crd/controller/controller.go [hunk 3]
```diff
@@ -90,7 +92,7 @@ func NewController(client *Client, options controller2.Options) model.ConfigStor
 	out := &controller{
 		client: client,
 		queue:  queue.NewQueue(1 * time.Second),
-		kinds:  make(map[string]*cacheHandler),
+		kinds:  make(map[resource.GroupVersionKind]*cacheHandler),
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crd/controller/controller.go]: Same here, assume we want to support k8s native gateway, it will not go to here

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crd/controller/controller.go]: Ignore, the gateway will not be native implement, it uses crd.

File: galley/pkg/config/schema/resource/schema.go (hunk 1 above)
Comment (howardjohn) [reply galley/pkg/config/schema/resource/schema.go]: yes it can, we read things from core v1 API group like pods

**Review by nmittler — COMMENTED** [review]
@howardjohn took a first pass ... some suggestions that may/may not make things a little easier.

File: galley/pkg/config/schema/collection/schemas.go [hunk 4]
```diff
@@ -108,25 +110,16 @@ func (s Schemas) MustFind(collection string) Schema {
...
+		if rs.Resource().GroupVersionKind() == gvk {
 			return rs, true
 		}
 	}
 
 	return nil, false
 }
 
-// MustFindByKind calls FindByKind and panics if not found.
-func (s Schemas) MustFindByKind(kind string) Schema {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply galley/pkg/config/schema/collection/schemas.go]: Yeah it was not used, so I removed it while I was there

[the remark this answers is not in the document]
Comment (howardjohn) [reply galley/pkg/config/schema/collection/schemas.go]: Ill add this back and remove the one that doesn't query for version as well

**Comment by howardjohn** [comment]
> Also, the `Kinds()` method should be changed to return GVK

There is one user of this, with a PR to remove that code completely open (the deprecated istioctl stuff). How about once that lands I remove it in a followup PR? I don't want to make this change to complex since its already large (and blocking me)

**Comment by nmittler** [comment]
@howardjohn 

> There is one user of this, with a PR to remove that code completely open (the deprecated istioctl stuff). How about once that lands I remove it in a followup PR? I don't want to make this change to complex since its already large (and blocking me)

SGTM

**Comment by howardjohn** [comment]
I cleaned up the last usage of finding with just Group and Kind, and added back a "MustFind"

**Review by liamawhite — APPROVED** [review]

**Review by incfly — APPROVED** [review]
/lgtm for only parts for security is at `pilot/pkg/secuity/authz` a helper on test func.

**Review by nmittler — APPROVED** [review]

**Review by rshriram — APPROVED** [review]

**Pull request merged**